ksqldb-execution/src/main/java/io/confluent/ksql/execution/util/ExpressionTypeManager.java

```diff
@@ -474,36 +484,90 @@ public Void visitFunctionCall(
-          }
+          final TypeContext childContext = expressionTypeContext.getCopy();
+          process(expression, childContext);
+          final SqlType resolvedArgType = childContext.getSqlType();
+          argTypesForFirstPass.add(SqlArgument.of(resolvedArgType));
         }
       }
 
-      final SqlType returnSchema = udfFactory.getFunction(argTypes).getReturnType(argTypes);
-      expressionTypeContext.setSqlType(returnSchema);
+      final KsqlScalarFunction function = udfFactory.getFunction(argTypesForFirstPass);
+
+      if (node.hasLambdaFunctionCallArguments()) {
+        final List<ParamType> paramTypes = function.parameters();
+        final Map<GenericType, SqlType> reservedGenerics = new HashMap<>();
+        final List<SqlArgument> argForReturnType = new ArrayList<>();
+        for (int i = 0; i < arguments.size(); i++) {
+          final Expression expression = arguments.get(i);
+          final ParamType parameter = paramTypes.get(i);
+          if (expression instanceof LambdaFunctionCall) {
+            final TypeContext childContext = expressionTypeContext.getCopy();
+
+            // the function returned from the UDF factory should have lambda
+            // at this index in the function arguments if there's a
+            // lambda node at this index in the function node argument list
+            if (!(parameter instanceof LambdaType)) {
+              throw new KsqlException("Error while processing lambda function. "
+                  + "This is most likely an internal error and a "
+                  + "Github issue should be filed for debugging.");
+            }
+
+            final LambdaType lambdaParameter = (LambdaType) parameter;
+            for (ParamType inputParam : lambdaParameter.inputTypes()) {
+              if (inputParam instanceof GenericType) {
+                final GenericType genericParam = (GenericType) inputParam;
+                if (!reservedGenerics.containsKey(genericParam)) {
+                  throw new KsqlException(
+                      String.format(
+                          "Could not verify type for generic %s.",
+                          genericParam.toString()));
+                }
+                childContext.addLambdaInputType(reservedGenerics.get(genericParam));
+              } else {
+                childContext.addLambdaInputType(
+                    SchemaConverters.functionToSqlConverter().toSqlType(inputParam));
+              }
+            }
+            final List<SqlType> sqlTypesForLambdaArgMapping =
+                new ArrayList<>(childContext.getLambdaInputTypes());
+            process(expression, childContext);
+
+            final SqlType resolvedArgType = childContext.getSqlType();
+            final SqlArgument lambdaArgument = SqlArgument.of(
+                SqlLambda.of(sqlTypesForLambdaArgMapping, resolvedArgType));
+
+            argForReturnType.add(lambdaArgument);
+            GenericsUtil.reserveGenerics(
```

Comment:
What's the scenario where there's a generic inside of a lambda that hasn't already been mapped?

Reply:
We wouldn't support something like this
```
  public <T,S, U> S randomFunction(
      @UdfParameter(description = "some array.") final List<T> list,
      @UdfParameter(description = "some state.") final S initialState,
      @UdfParameter(description = "some function.") final TriFunction<S, U, T, S> biFunction
  ) {
```
The LambdaType for the last argument would have Generics [S, U, T] in the input list and we wouldn't be able to handle U

Reply:
`select transform(array[1], (x,y) => 5) from hello emit changes;` would error out here

Reply:
Hmm yeah I see what you mean. That transform example isn't supported right now anyways, right? I guess it doesn't hurt to leave it in

Reply:
It's just a temporary example, and I don't think we need to support having unresolved generics in the lambda function in the near future.